transport/pool.go

```diff
@@ -0,0 +1,252 @@
+// When wanted connection is found it passes its index to PoolRefiller channel.
+func (c *ConnPool) Close(conn *Conn) {
+	for i, v := range c.conns {
+		if v.Load() == conn {
+			c.alertChan <- uint16(i)
+		}
+	}
+}
+
+// Handles errors and occasionally fills ConnPool.
+type PoolRefiller struct {
+	address    string
+	connConfig ConnConfig
+	connPool   ConnPool
+
+	activeShards uint16
+
+	connOptions      frame.ScyllaSupported
+	closeSinceRefill bool
+	currentDelay     time.Duration
+}
+
+// Fill happens in changing time intervals, these are its helper constants.
+const (
+	minFillBackoff        = 50 * time.Millisecond
+	maxFillBackoff        = 10 * time.Second
+	fillBackoffMultiplier = 2
+)
+
+// PoolRefiller endless function, for now it never shuts down.
+func (p *PoolRefiller) loop() {
+	runtime.LockOSThread()
+
+	p.currentDelay = minFillBackoff
+	ticker := time.NewTicker(p.currentDelay)
+	refilling := true
+
+	for {
+		// Wakes up either when it's refilling time or when there is connection to be closed.
+		select {
+		case <-ticker.C:
+			p.fill()
+			refilling = false
+		case idx := <-p.connPool.alertChan:
+			p.closeConn(idx)
+		}
+
+		if p.needsFilling() && !refilling {
+			// Time interval of refilling is adjusted based on what caused it.
+			if p.closeSinceRefill {
+				p.onClose()
+			} else {
+				p.currentDelay = minFillBackoff
+			}
+
+			ticker.Reset(p.currentDelay)
+			refilling = true
+		} else if !p.needsFilling() {
+			ticker.Stop()
+		}
+	}
+}
+
+// Actual connection closer.
+func (p *PoolRefiller) closeConn(idx uint16) {
+	conn := p.connPool.loadConn(idx)
+	if conn == nil {
+		return
+	}
+	conn.Close()
+	p.connPool.storeConn(idx, nil)
+	p.activeShards--
+	p.closeSinceRefill = true
+}
+
+// Time interval of refilling is adjusted here.
+func (p *PoolRefiller) onClose() {
+	p.closeSinceRefill = false
+	if maxFillBackoff > p.currentDelay*fillBackoffMultiplier {
+		p.currentDelay *= fillBackoffMultiplier
+	} else {
+		p.currentDelay = maxFillBackoff
+	}
+}
+
+// Check whether ConnPool is full of connections,
+// will be more complicated in the future.
+func (p *PoolRefiller) needsFilling() bool {
+	return p.activeShards < p.connPool.nrShards
+}
+
+// Opens first connection of ConnPool and fills read values from scyllaSupported.
+func (p *PoolRefiller) firstConnection() {
+	// First connection is needed to be done a bit differently,
+	// without opening it to a specific shard in order to read frame.ScyllaSupported.
+	conn, err := OpenConn(p.address, nil, p.connConfig)
+	if err != nil {
+		return
+	}
+	p.connOptions, err = getScyllaSupported(conn)
+	if err != nil {
+		return
+	}
+
+	p.connPool.nrShards = p.connOptions.NrShards
+	p.connPool.conns = make([]atomic.Value, p.connPool.nrShards)
+	p.connPool.storeConn(p.connOptions.Shard, conn)
+	p.connPool.msbIgnore = p.connOptions.MsbIgnore
+	p.connPool.alertChan = make(chan uint16, p.connPool.nrShards)
+	p.activeShards = 1
+}
+
+// Refills ConnPool.
+func (p *PoolRefiller) fill() {
```

Comment:
Can we move that just before loop and put other functions below in the order of dependency / appearance.

The objective is to be able to read it top down without too much jumping.